### PR TITLE
Add Enemy states chasing and roaming

### PR DESCRIPTION
**Pull request by ivan-BRHM**
#19 Adding features:
* Enemy roaming stage
* Enemy hp property, with a setter and a getter
* more?

### CODE REVIEW

File: my_sprites.py [hunk 1]
```diff
@@ -56,6 +56,14 @@ def __init__(
...
         return self._max_hp
 
+    @property
+    def hp(self):
+        return self._max_hp
+
+    @hp.setter
+    def hp(self, new_hp):
+        self._hp = min(new_hp, self.max_hp)
+
```
Comment (tobiasbp) [inline my_sprites.py:66]: One could set the _hp_ to _-10_!

Comment (tobiasbp) [reply my_sprites.py:76]: What about:
```
max(0, min(new_hp, self.max_hp))
```

**Review by tobiasbp — COMMENTED** [review]
Minor suggestion

File: my_sprites.py [hunk 2]
```diff
@@ -35,9 +39,13 @@ def __init__(
...
         # hp
         self._max_hp = max_hp
-        self.cur_hp = max_hp
+        self._hp = max_hp
 
+        self.window = window
         self.speed = speed
+        self.target = target
+        self.roaming_dist = roaming_dist
+        self.state = EnemyState.ROAMING
```
Comment (tobiasbp) [inline my_sprites.py:48]: Please add state to the arguments in _init()_. This state can be the default value.

File: my_sprites.py [hunk 3]
```diff
@@ -58,13 +66,23 @@ def __init__(
...
         return self._max_hp
 
+    @property
+    def hp(self):
+        return self._max_hp
+
+    @hp.setter
+    def hp(self, new_hp):
+        self._hp = max(0, min(new_hp, self.max_hp))
+
```
Comment (tobiasbp) [inline my_sprites.py:76]: Let's add a getter and a setter for the state. So, rename the state variable to _self._state_ internally.

File: my_sprites.py [hunk 4]
```diff
@@ -76,14 +94,38 @@ def go_to_position(self, target_pos: tuple[int, int]):
...
+        if self.state == EnemyState.CHASING:
+            self.path = []
+
+            angle_to_target = arcade.get_angle_radians(self.center_x, self.center_y, self.target.center_x, self.target.center_y)
+
+            self.center_x += math.sin(angle_to_target) * self.speed
+            self.center_y += math.cos(angle_to_target) * self.speed
+
+        # roaming state
+        if not self.path and self.state == EnemyState.ROAMING:
```
Comment (tobiasbp) [inline my_sprites.py:113]: Hmmm.... We should only look at the state... This makes it unclear what the ROAMING state does...

File: my_sprites.py (hunk 2 above)
Comment (tobiasbp) [inline my_sprites.py:48]: Please add state to the arguments in init(). This state can be the default value.

File: my_sprites.py [hunk 5]
```diff
@@ -76,14 +94,38 @@ def go_to_position(self, target_pos: tuple[int, int]):
 
     def on_update(self, delta_time: float = 1 / 60):
 
-        if not self.path:
+        # state control
+        if arcade.has_line_of_sight(self.position, self.target.position, self.barriers.blocking_sprites):
+            self.state = EnemyState.CHASING
+        else:
+            self.state = EnemyState.ROAMING
```
Comment (tobiasbp) [inline my_sprites.py:101]: When you have added a setter for _self.state_, like I asked about above, we can trigger emotes when the state changes here.

**Review by tobiasbp — CHANGES_REQUESTED** [review]
A few things missing for a complete implementation of states.

File: my_sprites.py (hunk 4 above)
Comment (ivan-BRHM) [reply my_sprites.py:113]: so... split up the if statements?

File: my_sprites.py (hunk 5 above)
Comment (ivan-BRHM) [reply my_sprites.py:101]: sure

File: my_sprites.py (hunk 4 above)
Comment (tobiasbp) [reply my_sprites.py:132]: Yes, I think so. To make it clear what happens in the _ROAMING_ state.

**Review by tobiasbp — CHANGES_REQUESTED** [review]
Cool feature.

**Review by tobiasbp — APPROVED** [review]
Looks good. Let's merge it in.

**Pull request merged**